lib/transfer.c

```diff
@@ -1804,12 +1804,12 @@ CURLcode Curl_retry_request(struct connectdata *conn,
   }
   if(retry) {
 #define CONN_MAX_RETRIES 5
-    if(conn->retrycount++ >= CONN_MAX_RETRIES) {
+    if(data->retrycount++ >= CONN_MAX_RETRIES) {
       failf(data, "Connection died, tried %d times before giving up",
             CONN_MAX_RETRIES);
```

Comment:
How about clearing `retrycount` again when it returns the failure, so that if the user tries the transfer again it'll start over with new attempts?

Reply:
Done. Thank you for the reminder

Reply:
@bagder While we are at it, should curl automatically retry a compile-time-number of times (or, should it retry at all?), in spite of the `--retry` option?

Reply:
This isn't really a "retry" in the `--retry` sense. This is a request that failed because the h2 server denied the request to be sent on that connection it was first issued on, so libcurl then tries on another connection. I think a (low) fixed retry count make sense here as I doubt anyone could come up with a better number if given the chance.

Reply:
@bagder I figured the `--retry` command line option is a tool thing, not a lib thing. But the lack of any retry-related options in the easy/multi API of libcurl was why I was under the impression that libcurl does not retry at all, and the handling of retry should be left to the program. That was just was my intuition on the design of the API, but apparently I was mistaken.

Reply:
Yes sure, it's not always crystal clear where to draw the line of how much libcurl itself should do and when it should simply fail and leave it to the app to retry. So far I've considered the "do not do anything more on this connection" to be more of a protocol advice and not an actual error to push to the application/user so I figured it would make sense to let libcurl do the retry there - returning on error for this case would be rather confusing to the user I fear... But this is not a very strong opinion of mine.